### PR TITLE
add web compiler option to pubspec.yaml, change --compiler to --web-compiler

### PR DESCRIPTION
**Pull request by jakemac53**
With the top level key in the pubspec being called `js-compiler`, it no longer makes sense for the command line arg to be just `compiler` (same argument holds about future compilers conflicting). Since `js-compiler` is really long I talked with @munificent and we landed on just `--js`.

Let me know if there are objections to changing the flag from `--compiler` to `--js`. If we could come up with a single character shorthand for `--js-compiler` that might be preferable, but I don't know of one that really makes sense.

cc @kevmoo @munificent @jwren @devoncarew 

EDIT: See comment below about naming changes

### CODE REVIEW

**Comment by jakemac53** [comment]
Ok, after some offline discussion I will update this to change the arg to `--web-compiler`, and I am going to change the pubspec config to have a top level `web` key, with `compiler` as a key under that, so it will look like this:

```yaml
web:
  compiler:
    debug: dartdevc
```

**Comment by jakemac53** [comment]
Updated now to reflect the new configuration.

**Comment by munificent** [comment]
👍  to the name change.

File: lib/src/command/barback.dart [hunk 1]
```diff
@@ -27,23 +27,28 @@ abstract class BarbackCommand extends PubCommand {
...
+      if (argResults.options.contains("web-compiler") &&
+          argResults.wasParsed("web-compiler")) {
         usageException(
-            "The --dart2js flag can't be used with the --compiler arg. "
-            "Prefer using the --compiler arg as --[no]-dart2js is deprecated.");
+            "The --dart2js flag can't be used with the --web-compiler arg. "
+            "Prefer using the --web-compiler arg as --[no]-dart2js is "
+            "deprecated.");
       }
       if (argResults['dart2js']) {
```
Comment (nex3) [inline lib/src/command/barback.dart]: Shouldn't this produce a warning as well?

File: lib/src/command/barback.dart [hunk 2]
```diff
@@ -27,23 +27,28 @@ abstract class BarbackCommand extends PubCommand {
...
         return Compiler.dart2JS;
       } else {
         return Compiler.none;
       }
-    } else if (argResults.options.contains("compiler")) {
-      return Compiler.byName(argResults["compiler"]);
+    } else if (argResults.options.contains("web-compiler") &&
+        argResults.wasParsed("web-compiler")) {
```
Comment (nex3) [inline lib/src/command/barback.dart]: Why not just `argResults["web-compiler"]`?

File: lib/src/pubspec.dart [hunk 3]
```diff
@@ -354,6 +355,57 @@ class Pubspec {
...
+        _error('"compiler" keys must be strings.', key.span);
+      }
+
+      final keyPattern = new RegExp(r"^[a-zA-Z0-9_-]+$");
+      if (!keyPattern.hasMatch(key.value)) {
+        _error(
+            '"compiler" keys may only contain letters, '
+            'numbers, hyphens and underscores.',
+            key.span);
+      }
```
Comment (nex3) [inline lib/src/pubspec.dart]: Why this restriction? As far as I remember, there aren't any such restrictions on Barback mode names.

File: lib/src/pubspec.dart [hunk 4]
```diff
@@ -354,6 +355,57 @@ class Pubspec {
 
   Map<String, String> _executables;
 
+  /// The settings for which web compiler to use in which mode.
+  ///
+  /// It is a map of strings to string. Each key is the name of a mode, and
+  /// the value is the name of the web compiler to use in that mode.
+  ///
+  /// Valid compiler values are all of [Compiler.names].
+  Map<String, String> get webCompiler {
```
Comment (nex3) [inline lib/src/pubspec.dart]: Why not `Map<String, Compiler>`?

File: test/pubspec_test.dart [hunk 5]
```diff
@@ -592,5 +592,61 @@ executables:
         expect(pubspec.executables['command'], equals('command'));
       });
     });
+
+    group("web", () {
+      test("can be empty", () {
+        var pubspec = new Pubspec.parse('web: {}', sources);
+        expect(pubspec.webCompiler, isEmpty);
+      });
```
Comment (nex3) [inline test/pubspec_test.dart]: Also can be null.

File: lib/src/command/barback.dart (hunk 1 above)
Comment (jakemac53) [inline lib/src/command/barback.dart]: done

File: lib/src/command/barback.dart (hunk 2 above)
Comment (jakemac53) [inline lib/src/command/barback.dart]: removed, there was a default value but I removed it as well.

File: lib/src/pubspec.dart (hunk 4 above)
Comment (jakemac53) [inline lib/src/pubspec.dart]: Oh sure, I like that, updated.

File: lib/src/pubspec.dart (hunk 3 above)
Comment (jakemac53) [inline lib/src/pubspec.dart]: Oh ok, I guess I was just assuming this, removed.

File: test/pubspec_test.dart (hunk 5 above)
Comment (jakemac53) [inline test/pubspec_test.dart]: I don't get a YamlMap in that case, but I added this to the "throws if not a map" test.

Comment (nex3) [inline test/pubspec_test.dart]: I believe in general we allow users to explicitly specify `null` for fields that are allowed to not exist, which means `web:` without anything after it is valid.

File: lib/src/command/barback.dart [hunk 6]
```diff
@@ -35,20 +35,20 @@ abstract class BarbackCommand extends PubCommand {
...
       }
-      if (argResults['dart2js']) {
+      if (argResults["dart2js"]) {
         return Compiler.dart2JS;
       } else {
         return Compiler.none;
       }
-    } else if (argResults.options.contains("web-compiler") &&
-        argResults.wasParsed("web-compiler")) {
+    } else if (argResults.options.contains("web-compiler")) {
```
Comment (nex3) [inline lib/src/command/barback.dart]: I'm still not quite sure why this isn't just `argResults["web-compiler"]` with a default value of `false`.

File: test/pubspec_test.dart (hunk 5 above)
Comment (jakemac53) [inline test/pubspec_test.dart]: ok, done. I also added a test for a null compiler value

File: lib/src/command/barback.dart (hunk 6 above)
Comment (jakemac53) [inline lib/src/command/barback.dart]: It doesn't feel right to have a default value that isn't a valid compiler name to me. If you want I can do that though.

Comment (nex3) [inline lib/src/command/barback.dart]: Oh I see, I was thinking this was a boolean value for some reason.

**Comment by kevmoo** [comment]
Cherry pick request to sdk?

From mobile. Sorry for brevity and/or spelling.

**Comment by jakemac53** [comment]
cl out to merge into sdk here https://codereview.chromium.org/2908783003/ then I will send out the cherry pick request

**Comment by jakemac53** [comment]
https://github.com/dart-lang/sdk/issues/29741
